packages/react-components/src/DemocracyLocks.tsx

```diff
@@ -20,50 +20,101 @@ interface Props {
   value?: DeriveDemocracyLock[];
 }
 
+interface Entry {
+  details: React.ReactNode;
+  headers: React.ReactNode[];
+  isCountdown: boolean;
+  isFinished: boolean;
+}
+
+interface State {
+  maxBalance: BN;
+  sorted: Entry[];
+}
+
 let id = 0;
 
+// group by header & details
+//   - all unlockable together
+//   - all ongoing together
+//   - unlcokes are displayed individually
+function groupLocks (t: <T = string> (key: string, options?: { replace?: Record<string, string> }) => T, bestNumber: BN, locks: DeriveDemocracyLock[] = []): State {
+  return {
+    maxBalance: bnMax(...locks.map(({ balance }) => balance)),
+    sorted: locks
+      .map((info): [DeriveDemocracyLock, BN] => [info, info.unlockAt.gt(bestNumber) ? info.unlockAt.sub(bestNumber) : BN_ZERO])
+      .sort((a, b) => a[0].referendumId.cmp(b[0].referendumId))
+      .sort((a, b) => a[1].cmp(b[1]))
+      .sort((a, b) => a[0].isFinished === b[0].isFinished ? 0 : (a[0].isFinished ? -1 : 1))
+      .reduce((sorted: Entry[], [{ balance, isFinished, referendumId, vote }, blocks]): Entry[] => {
+        const isCountdown = blocks.gt(BN_ZERO);
+        const header = <div>#{referendumId.toString()} {formatBalance(balance, { forceUnit: '-' })} {vote.conviction.toString()}</div>;
+        const prev = sorted.length ? sorted[sorted.length - 1] : null;
+
+        if (!prev || (isCountdown || (isFinished !== prev.isFinished))) {
+          sorted.push({
+            details: (
+              <div className='faded'>
+                {isCountdown
+                  ? (
+                    <BlockToTime
+                      blocks={blocks}
+                      label={`${t<string>('{{blocks}} blocks', { replace: { blocks: formatNumber(blocks) } })}, `}
+                    />
+                  )
+                  : isFinished
+                    ? t<string>('lock expired')
+                    : t<string>('ongoing referendum')
+                }
+              </div>
+            ),
+            headers: [header],
+            isCountdown,
+            isFinished
+          });
+        } else {
+          prev.headers.push(header);
+        }
+
+        return sorted;
+      }, [])
+  };
+}
+
 function DemocracyLocks ({ className = '', value }: Props): React.ReactElement<Props> | null {
   const { t } = useTranslation();
   const { api } = useApi();
   const bestNumber = useCall<BN>(api.derive.chain.bestNumber, []);
   const [trigger] = useState(`${Date.now()}-democracy-locks-${++id}`);
```

Comment:
unrelated: what is the benefit of doing this rather than using a `const`?

Reply:
The idea is to not change the state object when the structure hasn't changed - when this comment was made that check was not in yet.